### PR TITLE
Implement transaction management in BulkReplyResolveService

### PR DESCRIPTION
**Pull request by tonyblank**
- Introduced a comprehensive transaction management system to handle bulk reply and resolve operations.
- Added rollback capabilities to ensure data consistency during failures.
- Enhanced the service to support atomic operations, allowing all operations to succeed or fail together.
- Implemented checkpointing for better transaction tracking and recovery.
- Updated the service to include detailed transaction logging and audit reporting.
- Added unit tests to cover various scenarios, including success, failure, and edge cases.

## Summary by CodeRabbit

- **New Features**
  - Introduced robust transaction management for bulk reply and resolve operations, including configurable rollback strategies, checkpointing, and detailed audit reporting.
  - Added rollback handlers for reply and resolve actions, supporting automatic undo of thread resolution and comprehensive error logging.
  - Enhanced bulk operation summaries with transaction metadata and audit details.

- **Bug Fixes**
  - Improved error handling for transaction failures with new error codes.

- **Tests**
  - Added extensive unit tests for transaction management, rollback handlers, and enhanced bulk reply/resolve workflows to ensure reliability and correct behavior.

### CODE REVIEW

File: src/toady/bulk_reply_resolve_service.py [hunk 1]
```diff
@@ -60,26 +71,54 @@ def __init__(
...
+        )
+        self.transaction_manager.register_rollback_handler(
+            OperationType.REPLY_POST, rollback_handler
+        )
+        self.transaction_manager.register_rollback_handler(
+            OperationType.THREAD_RESOLVE, rollback_handler
+        )
+        self.transaction_manager.register_rollback_handler(
+            OperationType.THREAD_UNRESOLVE, rollback_handler
+        )
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply src/toady/bulk_reply_resolve_service.py:112]: Great catch\! I've refactored the repetitive rollback handler registration to use a loop.

**What I changed:**
- Replaced three separate register_rollback_handler calls with a single loop
- Used a list of OperationType values to iterate over
- Maintained the same functionality with cleaner, more maintainable code

**Why this approach is better:**
1. **DRY Principle**: Eliminates code duplication - if we need to register more operation types or change the handler, we only modify one place
2. **Readability**: The intent is clearer - we're registering the same handler for multiple operation types
3. **Maintainability**: Adding new operation types or changing the registration logic requires minimal code changes
4. **Consistency**: Follows Python idioms for reducing repetition

The refactored code maintains identical functionality while being more concise and maintainable. This is a common pattern when you need to apply the same operation to multiple items.

File: src/toady/bulk_reply_resolve_service.py [hunk 2]
```diff
@@ -117,24 +158,56 @@
...
+                assert transaction_id is not None  # Should be set for non-dry-run
+                return self._perform_atomic_operations_with_transaction(
+                    operations, transaction_id
+                )
             else:
-                return self._perform_non_atomic_operations(operations)
+                assert transaction_id is not None  # Should be set for non-dry-run
+                return self._perform_non_atomic_operations_with_transaction(
+                    operations, transaction_id
+                )
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply src/toady/bulk_reply_resolve_service.py:185]: Absolutely right\! Using assert statements for production validation is a serious bug waiting to happen.

**What I changed:**
- Replaced both assert statements with explicit BulkOperationError exceptions
- Added descriptive error messages and context for debugging
- Used our custom exception type that matches the existing error handling pattern

**Why assert statements are dangerous in production:**
1. **Python Optimization Removal**: When Python runs with -O or -OO flags, assert statements are completely stripped out, making these critical checks disappear in production
2. **Silent Failures**: Without the check, transaction_id could be None, leading to confusing TypeErrors or AttributeErrors later
3. **Poor Error Messages**: AssertionError doesn't provide meaningful context about what went wrong

**Why BulkOperationError is better:**
1. **Always Active**: Custom exceptions work regardless of Python optimization flags
2. **Rich Context**: We include the atomic/dry_run state to help diagnose the root cause
3. **Consistent Error Handling**: Uses the same exception type as other bulk operation failures
4. **Clear Intent**: Makes it obvious this is a business logic violation, not a debug assertion

This change ensures robust error handling that works reliably in all environments, from development to production deployments.

File: src/toady/transaction_manager.py [hunk 3]
```diff
@@ -0,0 +1,601 @@
...
+                self.logger.warning(f"Best effort rollback failed: {e}")
+                rollback_success = False
+
+        # For abort, always mark as FAILED regardless of rollback success
+        # Abort indicates the transaction was terminated due to an error
+        self._current_transaction.status = TransactionStatus.FAILED
+        self._current_transaction.error_message = error_message
+        if self._current_transaction.end_time is None:
+            self._current_transaction.end_time = datetime.now()
+
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply src/toady/transaction_manager.py:435]: Spot on\! This is exactly the bug we fixed earlier re-surfacing. You've caught a critical regression that would break transaction status accuracy.

**The Problem:**
- rollback_transaction() correctly sets status to ROLLED_BACK when successful
- abort_transaction() then unconditionally overwrites it with FAILED
- This destroys the accurate status information, breaking monitoring and debugging

**The Fix:**
I've implemented the suggested solution to only set FAILED status when transaction is still ACTIVE, preserving the ROLLED_BACK status when rollback succeeded.

**Why This Matters:**
- **Status Accuracy**: Preserves the semantic difference between "rolled back cleanly" vs "failed to abort"
- **Monitoring Integration**: External systems can distinguish between different failure modes
- **Debugging Value**: Developers can see that rollback succeeded even though transaction was aborted
- **Semantic Correctness**: A successfully rolled back transaction should show ROLLED_BACK status

**Transaction State Semantics:**
- ACTIVE → FAILED: Abort without successful rollback
- ACTIVE → ROLLED_BACK: Rollback succeeded during abort  
- ROLLED_BACK → ROLLED_BACK: Preserve successful rollback status

This maintains status integrity while still recording the error message and ensuring proper transaction lifecycle management.

File: src/toady/bulk_reply_resolve_service.py [hunk 4]
```diff
@@ -240,6 +320,288 @@
...
+                                rollback_success=True,  # Rollback performed by abort
+                            )
+                            failed_results.append(failed_result)
+
+                        return BulkOperationSummary(
+                            total_operations=len(operations),
+                            successful_operations=0,
+                            failed_operations=len(operations),
+                            results=failed_results + [result],
+                            atomic_failure=True,
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply src/toady/bulk_reply_resolve_service.py:389]: Excellent catch! The results list was incomplete for atomic failures.

**The Problem:**
failed_operations=len(operations) claimed ALL operations failed but results list only contained attempted operations, creating a mismatch.

**The Solution:**
Created complete results list by adding placeholder BulkOperationResult objects for unattempted operations with error message indicating they were not attempted due to atomic failure.

**Benefits:**
- Complete accountability - every operation has a result entry
- Clear distinction between rolled-back vs unattempted operations
- Metrics accuracy - len(results) equals total_operations
- Better debugging - can see which operations were attempted vs skipped

This maintains atomic failure semantics while providing complete traceability for downstream metrics.

File: src/toady/transaction_manager.py [hunk 5]
```diff
@@ -0,0 +1,602 @@
...
+            TransactionStatus.ROLLED_BACK if success else TransactionStatus.FAILED
+        )
+        self._current_transaction.end_time = datetime.now()
+
+        self.logger.info(
+            f"Transaction {self._current_transaction.transaction_id} rollback: "
+            f"{'successful' if success else 'failed'}"
+        )
+        return success
+
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply src/toady/transaction_manager.py:370]: Critical observation! Leaving completed transactions dangling creates serious lifecycle management issues.

**The Problem:**
After rollback_transaction(), the transaction had end_time set and status marked as ROLLED_BACK/FAILED, but remained in _current_transaction. This allows:
- Accidentally working on a finished transaction
- Potential to commit a rolled-back transaction
- Memory leaks from uncleaned transaction state

**The Solution:**
Added proper transaction lifecycle management:
1. Move completed transaction to _transaction_history
2. Trim history if it exceeds max_operation_history
3. Clear _current_transaction to None

**Why This Matters:**
- **Lifecycle Integrity**: Ensures finished transactions cannot be reused
- **Memory Management**: Prevents accumulation of stale transaction references
- **State Consistency**: Clear separation between active and historical transactions
- **Audit Trail**: Rolled-back transactions properly recorded in history

This follows the same pattern as commit_transaction() and abort_transaction(), ensuring consistent transaction lifecycle management across all completion paths.

File: src/toady/bulk_reply_resolve_service.py [hunk 6]
```diff
@@ -240,6 +320,304 @@ def _create_bulk_operations(
...
+                        successful_operations=0,
+                        failed_operations=len(operations),
+                        results=successful_results + [failed_result],
+                        atomic_failure=True,
+                        rollback_performed=True,
+                        transaction_id=transaction_id,
+                        transaction_status=TransactionStatus.FAILED.value,
+                        checkpoints_created=checkpoints_created,
+                        audit_report=audit_report,
+                    )
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply src/toady/bulk_reply_resolve_service.py:441]: Perfect catch! You found the duplicate hard-coded status I missed in the exception handler.

**The Issue:**
Second BulkOperationSummary return still used TransactionStatus.FAILED.value instead of actual audit report status.

**The Fix:**
Applied same pattern with defensive null checking:
- Extract status from audit_report.get(\status\) 
- Fallback to FAILED.value if audit_report is None
- Maintains consistency across all code paths

**Result:**
All transaction status reporting now uses authoritative source from transaction manager, ensuring summary and audit report always align.

File: src/toady/transaction_manager.py [hunk 7]
```diff
@@ -0,0 +1,610 @@
...
+            self._current_transaction.status = TransactionStatus.FAILED
+        self._current_transaction.error_message = error_message
+        if self._current_transaction.end_time is None:
+            self._current_transaction.end_time = datetime.now()
+
+        # Move to history
+        self._transaction_history.append(self._current_transaction)
+        transaction_id = self._current_transaction.transaction_id
+        self._current_transaction = None
+
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply src/toady/transaction_manager.py:441]: Excellent catch! You identified a memory leak in the abort path - missing history trimming.

**The Problem:**
- commit_transaction() properly trims _transaction_history to max_operation_history
- abort_transaction() was missing the same trimming logic  
- High-frequency aborted transactions could cause unbounded memory growth

**The Solution:**
Added identical trimming logic to abort_transaction() that matches commit_transaction().

**Why This Matters:**
- **Memory Safety**: Prevents slow memory leaks in high-volume scenarios
- **Consistency**: All terminal operations now have same memory management
- **Production Reliability**: Critical for long-running services with many transactions

**Pattern Recognition:**
This is a common oversight - implementing cleanup in one code path but missing it in error paths. Good reminder to audit all similar code paths for consistency.

File: src/toady/transaction_manager.py [hunk 8]
```diff
@@ -0,0 +1,610 @@
...
+            raise TransactionError("No active transaction to rollback")
+
+        operations_to_rollback = list(reversed(self._current_transaction.operations))
+        success = self._perform_rollback(operations_to_rollback)
+
+        self._current_transaction.status = (
+            TransactionStatus.ROLLED_BACK if success else TransactionStatus.FAILED
+        )
+        self._current_transaction.end_time = datetime.now()
+
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply src/toady/transaction_manager.py:364]: Brilliant observation about transaction semantics! Setting end_time in non-terminal operations creates incorrect audit data.

**The Problem:**
- rollback_transaction() is non-terminal - transaction remains active
- Setting end_time freezes duration calculation prematurely  
- Later commit_transaction() or abort_transaction() would overwrite timestamp
- Makes elapsed-time metrics meaningless and audit data confusing

**The Solution:**
Removed end_time assignment from rollback_transaction(), leaving it for terminal operations only.

**Semantic Correctness:**
- **rollback_transaction()**: Changes status but keeps transaction alive
- **commit_transaction()**: Terminal operation, sets end_time  
- **abort_transaction()**: Terminal operation, sets end_time

**Why This Matters:**
- **Accurate Auditing**: Duration reflects actual transaction lifetime
- **Clear Semantics**: Non-terminal vs terminal operations properly distinguished
- **Reliable Metrics**: Elapsed time calculations remain meaningful

This ensures audit timestamps accurately reflect the true transaction lifecycle.

File: tests/unit/test_transaction_manager.py [hunk 9]
```diff
@@ -0,0 +1,498 @@
...
+        """Test registering rollback handlers."""
+        transaction_manager.register_rollback_handler(
+            OperationType.REPLY_POST, mock_rollback_handler
+        )
+
+        assert OperationType.REPLY_POST in transaction_manager._rollback_handlers
+        assert (
+            transaction_manager._rollback_handlers[OperationType.REPLY_POST]
+            is mock_rollback_handler
+        )
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply tests/unit/test_transaction_manager.py:83]: Absolutely right\! Accessing private attributes makes tests brittle and prone to breaking during refactoring.

**The Problem:**
Direct access to _rollback_handlers couples test to internal implementation details. Private attributes (underscore prefix) are implementation details that can change without notice.

**The Solution:**
1. Added public get_rollback_handler() method to TransactionManager
2. Updated test to use public API:
   - get_rollback_handler_count() == 1
   - get_rollback_handler(OperationType.REPLY_POST) is mock_rollback_handler

**Why This Pattern:**
- **Encapsulation**: Tests use stable public interface
- **Maintainability**: Tests won't break from internal refactoring
- **Documentation**: Public methods express intended test behavior
- **Future-proof**: Implementation can change without test updates

**Best Practice:**
Tests should only access public APIs. If tests need to verify internal state, that's a signal to add appropriate public accessors that express the business intent being tested.

File: tests/unit/services/test_bulk_reply_resolve_service_enhanced.py [hunk 10]
```diff
@@ -0,0 +1,523 @@
...
+            reply_body="test message",
+            operation_id="op_456",
+        )
+
+        # Mock successful services
+        bulk_service.reply_service.post_reply.return_value = {"reply_id": "reply_789"}
+        bulk_service.resolve_service.resolve_thread.return_value = {"success": True}
+
+        result = bulk_service._execute_single_operation_with_transaction(operation)
+
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply tests/unit/services/test_bulk_reply_resolve_service_enhanced.py:322]: Thanks for the vigilance! However, this issue was already resolved in a previous iteration.

**Current State:**
The method signature was updated to remove the unused _transaction_id parameter:
- Method: _execute_single_operation_with_transaction(self, operation)
- Test call: _execute_single_operation_with_transaction(operation)

**What Happened:**
In the previous round of feedback, I removed the unused _transaction_id parameter from both the method signature and all call sites, including test calls.

**Verification:**
The tests are currently passing, confirming the signatures match correctly.

**Key Learning:**
This highlights the importance of running tests after each change to catch signature mismatches immediately. The automated CI pipeline catches these issues quickly, preventing them from reaching review.

File: src/toady/bulk_reply_resolve_service.py [hunk 11]
```diff
@@ -240,6 +320,304 @@ def _create_bulk_operations(
...
+                            failed_result = BulkOperationResult(
+                                operation_id=prev_result.operation_id,
+                                thread_id=prev_result.thread_id,
+                                success=False,  # All operations fail in atomic mode
+                                reply_result=prev_result.reply_result,
+                                resolve_result=prev_result.resolve_result,
+                                error="Rolled back due to atomic failure",
+                                rollback_attempted=True,
+                                rollback_success=True,  # Rollback performed by abort
+                            )
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply src/toady/bulk_reply_resolve_service.py:380]: Excellent catch! Hard-coding rollback_success=True ignores actual rollback outcome.

**The Problem:**
abort_transaction() returns boolean for rollback success but we hard-coded True, providing misleading status when rollback fails.

**The Solution:**
Capture actual result: rollback_ok = abort_transaction(...) and use rollback_success=rollback_ok

**Why Critical:**
- **Accuracy**: Status reflects reality, not assumptions
- **Debugging**: Failed rollbacks properly identified  
- **Trust**: System provides reliable diagnostic information

Never assume operations succeed - always capture and propagate actual results, especially in failure paths.

File: src/toady/transaction_manager.py [hunk 12]
```diff
@@ -0,0 +1,614 @@
...
+        operation_id = str(uuid.uuid4())
+        operation = OperationRecord(
+            operation_id=operation_id,
+            operation_type=operation_type,
+            timestamp=datetime.now(),
+            thread_id=thread_id,
+            data=data.copy(),
+            rollback_data=rollback_data.copy() if rollback_data else None,
+        )
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply src/toady/transaction_manager.py:236]: Great point about shallow vs deep copying\! Shallow copy can cause subtle bugs with nested data structures.

**The Problem:**
- rollback_data.copy() only clones top level
- Nested dicts/lists remain shared between original and copy
- Mutations to nested objects affect both copies unexpectedly

**The Solution:**
1. Import copy module
2. Use copy.deepcopy(rollback_data) for full isolation
3. Apply same pattern to data parameter for consistency

**Why Deep Copy Matters:**
- **Isolation**: Complete separation between original and copy
- **Safety**: Prevents unintended side effects from mutations
- **Predictability**: Changes to one don't affect the other
- **Debugging**: Eliminates mysterious shared-state bugs

**Best Practice:**
Use deep copy for complex data structures in transaction systems where data integrity is critical. The performance cost is typically negligible compared to the safety benefit.

File: src/toady/bulk_reply_resolve_service.py [hunk 13]
```diff
@@ -240,6 +320,304 @@ def _create_bulk_operations(
...
+                        return BulkOperationSummary(
+                            total_operations=len(operations),
+                            successful_operations=0,
+                            failed_operations=len(operations),
+                            results=all_failed_results,
+                            atomic_failure=True,
+                            rollback_performed=True,
+                            transaction_id=transaction_id,
+                            transaction_status=audit_report["status"],
+                            checkpoints_created=checkpoints_created,
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply src/toady/bulk_reply_resolve_service.py:407]: Excellent catch! Hard-coding rollback_performed=True creates misleading metrics.

**The Problem:**
We captured rollback_ok from abort_transaction() but ignored it in the summary, always reporting rollback_performed=True regardless of actual outcome.

**The Fix:**
Changed rollback_performed=True to rollback_performed=rollback_ok

**Why This Matters:**
- **Accurate Metrics**: Summary now reflects reality, not assumptions
- **Debugging**: Failed rollbacks are properly reported to callers
- **Monitoring**: Downstream systems get honest rollback success rates
- **Trust**: No false positives in operational dashboards

**Learning Point:**
Always propagate actual operation results through all layers. Hard-coded success flags are lies waiting to bite you during incident response when you desperately need accurate status information.

File: src/toady/transaction_manager.py [hunk 14]
```diff
@@ -0,0 +1,628 @@
...
+
+        checkpoint_id = str(uuid.uuid4())
+        checkpoint = Checkpoint(
+            checkpoint_id=checkpoint_id,
+            timestamp=datetime.now(),
+            operation_count=len(self._current_transaction.operations),
+            description=description,
+            data=data.copy() if data else {},
+        )
+
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply src/toady/transaction_manager.py:285]: Absolutely right\! Shallow copy in checkpoint data is a mutation hazard.

**The Problem:**
data.copy() only clones the first level of the dictionary. Nested dicts/lists are still shared references, allowing unintended mutations to corrupt checkpoint data.

**The Fix:**
Changed data.copy() to copy.deepcopy(data) for full isolation

**Example of the Hazard:**
```python
# With shallow copy - BAD
original = {'metrics': {'count': 5}}
checkpoint = {'data': original.copy()}
original['metrics']['count'] = 999  # Corrupts checkpoint\!

# With deep copy - GOOD  
checkpoint = {'data': copy.deepcopy(original)}
original['metrics']['count'] = 999  # Checkpoint preserved
```

**Why Deep Copy for Checkpoints:**
- **Immutability**: Checkpoint state never changes after creation
- **Recovery**: Rollback data must be pristine and untouchable
- **Debugging**: Historical state stays accurate for forensics

**Best Practice:**
Use deep copy whenever storing state that must remain immutable, especially in transaction/checkpoint systems where data integrity is critical.

File: src/toady/bulk_reply_resolve_service.py [hunk 15]
```diff
@@ -240,6 +320,304 @@ def _create_bulk_operations(
...
+                    )
+
+                    return BulkOperationSummary(
+                        total_operations=len(operations),
+                        successful_operations=0,
+                        failed_operations=len(operations),
+                        results=successful_results + [failed_result],
+                        atomic_failure=True,
+                        rollback_performed=True,
+                        transaction_id=transaction_id,
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply src/toady/bulk_reply_resolve_service.py:435]: Excellent follow-up\! You caught the identical bug in the exception handling path.

**The Problem:**
I fixed the main atomic failure branch but missed the exception path that also hard-coded rollback_performed=True, creating the same misleading status reporting.

**The Fix:**
1. **Captured actual rollback result**: rollback_ok = abort_transaction(...)
2. **Updated previous results**: Set rollback_attempted/rollback_success on earlier successful operations
3. **Honest summary**: Use rollback_performed=rollback_ok instead of True

**Why Update Previous Results:**
When an exception triggers rollback, previously successful operations in the transaction get rolled back too. Their rollback status must reflect the actual outcome:
```python
for res in successful_results:
    res.rollback_attempted = True
    res.rollback_success = rollback_ok  # Actual outcome, not assumption
```

**Critical Learning:**
Error handling paths are where truth goes to die. It's tempting to assume operations failed completely, but in atomic transactions, rollback success/failure is independent of the original operation failure. Always capture and propagate actual rollback outcomes - your future incident response self will thank you when you need accurate diagnostics.

File: src/toady/bulk_reply_resolve_service.py [hunk 16]
```diff
@@ -117,24 +157,64 @@ def bulk_reply_and_resolve(
...
             # Execute operations
             if dry_run:
                 return self._perform_dry_run(operations)
             elif atomic:
-                return self._perform_atomic_operations(operations)
+                if transaction_id is None:
+                    raise BulkOperationError(
+                        message="Transaction ID not initialized for atomic op",
+                        context={"atomic": atomic, "dry_run": dry_run},
+                    )
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply src/toady/bulk_reply_resolve_service.py:180]: Great refactoring suggestion\! You're absolutely right about the unnecessary elif.

**The Problem:**
Using `elif atomic:` after an early return creates dead weight in the control flow. The `if dry_run:` block returns early, so execution never reaches the elif unless dry_run is False.

**The Fix:**
Changed `elif atomic:` to `if atomic:` since the previous condition returns early.

**Why This Improves Code Quality:**
1. **Reduced Cyclomatic Complexity**: Fewer branching paths to consider
2. **Cleaner Logic**: Each condition stands independently  
3. **Linter Compliance**: Eliminates R1705 'unnecessary-else-after-return' warning
4. **Readability**: No need to mentally track the relationship between conditions

**Code Pattern:**
```python
# Before: Unnecessary elif after early return
if condition_a:
    return early_result()
elif condition_b:  # Dead weight - condition_a is always False here
    do_something()

# After: Clean independent conditions  
if condition_a:
    return early_result()
if condition_b:  # Clear intent - check condition_b independently
    do_something()
```

**Learning Point:**
When a branch returns early, subsequent conditions don't need elif - they can be independent if statements. This makes code more maintainable and reduces cognitive load.

File: src/toady/bulk_reply_resolve_service.py [hunk 17]
```diff
@@ -240,6 +320,310 @@ def _create_bulk_operations(
...
+                    # Abort and capture actual rollback outcome
+                    rollback_ok = self.transaction_manager.abort_transaction(
+                        f"Exception in operation {operation.operation_id}: {str(e)}"
+                    )
+
+                    # Make earlier successes honest about rollback
+                    for res in successful_results:
+                        res.rollback_attempted = True
+                        res.rollback_success = rollback_ok
+
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply src/toady/bulk_reply_resolve_service.py:429]: Excellent catch\! The failed operation's rollback status was incorrectly reported.

**The Problem:**
In the exception handling path, we updated rollback flags for previous successful operations but forgot to update the failed operation itself, leaving rollback_attempted=False despite abort_transaction() being called.

**The Fix:**
Added rollback flag updates for the failed operation:
```python
# Also mark the failed operation itself as rolled back
failed_result.rollback_attempted = True
failed_result.rollback_success = rollback_ok
```

**Why This Matters:**
- **Complete Accuracy**: All operations in atomic transactions get rolled back together
- **Consistent Reporting**: Failed operation shows same rollback status as successful ones
- **Debugging Value**: Clear record that rollback was attempted for this operation too

**Learning Point:**
When an exception triggers transaction rollback, ALL operations in the transaction are affected - both the ones that succeeded AND the one that failed. The rollback status should reflect this reality for every operation, ensuring no operation is left with misleading default values that suggest rollback was never attempted.
